src/pybind/mgr/mgr_util.py

```diff
@@ -164,3 +166,65 @@ def verify_tls_files(cert_fname, pkey_fname):
         logger.warning(
             'Private key {} and certificate {} do not match up: {}'.format(
                 pkey_fname, cert_fname, str(e)))
+
+
+def get_rates_from_data(data, use_abs=None):
+    """ Rates from time series data
+
+    :param data: Time series data
+    :type data: list[tuple[int, float]]
+
+    :param use_abs: If the derivative returned should be an absolute one
+    :type use_abs: bool (default None)
+        Using the absolute value is a temporary workaround for the pg decrease issue, that will
+        reduce the maximum amount in all time series data of a pool. Normally the maximum amount
+        should always increase.
+
+    :return: The derivative between all time series data points
+    :rtype: list[tuple[int, float]]
+
+    >>> get_rates_from_data([])
+    []
+    >>> get_rates_from_data([[1, 42]])
+    []
+    >>> get_rates_from_data([[0, 0], [2, 1], [3, 0], [4, -2]])
+    [(2, 0.5), (3, -1.0), (4, -2.0)]
+    >>> get_rates_from_data([[0, 0], [2, 1], [3, 0], [4, -2]], True)
+    [(2, 0.5), (3, 1.0), (4, 2.0)]
+    """
+    if not data or len(data) == 1:
+        return []
+    return [(data2[0], derivative(data1, data2, use_abs)) for data1, data2 in pairwise(data)]
+
+
+def derivative(p1, p2, use_abs=False):
+    """ Derivative between two time series data points
+
+    :param p1: Time series data point
+    :type p1: tuple[int, float]
+    :param p2: Time series data point
+    :type p2: tuple[int, float]
+
+    :param use_abs: If the derivative returned should be an absolute one
+    :type use_abs: bool (default False)
+        Using the absolute value is a temporary workaround for the pg decrease issue, that will
+        reduce the maximum amount in all time series data of a pool. Normally the maximum amount
+        should always increase.
+
+    :return: Derivative between both points
+    >>> values = [0, 1]
+    >>> times = [0, 0]
+    >>> derivative(*zip(times, values))
+    1.0
+    >>> times = [0, 2]
+    >>> derivative(*zip(times, values))
+    0.5
+    >>> values = [1, 0]
+    >>> derivative(*zip(times, values))
+    -0.5
+    >>> derivative(*zip(times, values), True)
+    0.5
+    """
```

Comment:
Nice unit test coverage. What about adding edge cases:
- `t=[0,0]`, `v=[0,0]` -> ZeroDivisionError -> NotEnoughTimeHasPassedError?
- `t=[0,0]`, `v=[0, float('nan')]` -> nan 
- `t=[0,float('inf')]`, `v=[0, 3]` -> 0
- `t=[0,float('inf')]`, `v=[0, float('inf')]` -> nan

Reply:
I'd prefer to have the scope of the doc tests limited. Doc tests are not the best place for extensive edge case testing.

Reply:
Why not? For simple functions, like math operations, they become both documentation and tests. And edge cases are the things you definitely want to include in a unit test, because otherwise you may have undefined behaviours.

Reply:
> Why not?

Because verbosely documenting edge cases rarely improve the readability of docstrings. I'd argue for keeping the number of tests within docstrings limited to what actually shows non-trivial behaviour.

Reply:
I personally find that doctests are documentation that becomes tested (for free). If @Devp00l adds a comment explaining this function's quirks: "when 2 samples refer to the same timestamp this returns the rate as if they were 1 second separated". If someone else later changes the behaviour and they somehow forget to amend that documentation accordingly (which is more common that it seems), there we have mismatching docs.

Reply:
Just keep in mind the doc string is already extraordinary long. At some point we have to evaluate how much documentation we want for purely internal functions.

Reply:
I will only add the case for the same timestamp, which will return 0 and should not return a failure or just skip the value.

Reply:
> I will only add the case for the same timestamp, which will return 0 and should not return a failure or just skip the value.

Then I'm afraid that I haven't made myself understood. The "same-timestamp-case" should never return 0, as `math.log(0)` never returns anything except `ValueError: math domain error`.